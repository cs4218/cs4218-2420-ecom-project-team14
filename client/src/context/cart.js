import axios from "axios";
import { createContext, useContext, useEffect, useReducer } from "react";
import toast from "react-hot-toast";
import { useAuth } from "./auth";

const CartContext = createContext();

const cartReducer = (state, action) => {
  switch (action.type) {
    case "ADD_TO_CART": {
      const { slug, price, productId } = action.payload;
      toast.success("Add to Cart Successfully");
      return {
        ...state,
        [slug]: state[slug]
          ? { quantity: state[slug].quantity + 1, price, productId}
          : { quantity: 1, price, productId},
      };
    }
    case "REMOVE_FROM_CART": {
      const newState = { ...state };
      delete newState[action.payload.slug];
      toast.success("Remove from Cart Successfully");
      return newState;
    }
    case "UPDATE_QUANTITY": {
      const { slug, quantity, price, productId } = action.payload;
      if (quantity <= 0) {
        const newState = { ...state };
        delete newState[slug];
        return newState;
      }
      toast.success("Update Cart Quantity Successfully");
      return {
        ...state,
        [slug]: { ...state[slug], quantity, price, productId },
      };
    }
    case "CLEAR_CART":
      toast.success("Cart Cleared Successfully");
      return {};
    default:
      return state;
  }
};

export const CartProvider = ({ children }) => {
  const auth = useAuth();

  const user = auth.user ?? "guest";

  const [cart, dispatch] = useReducer(cartReducer, {}, () => {
    const localData = localStorage.getItem(`cart-${user}`);
    return localData ? JSON.parse(localData) : {};
  });

  useEffect(() => {
    localStorage.setItem(`cart-${user}`, JSON.stringify(cart));
  }, [cart]);

  const addToCart = async (slug) => {
    const { data } = await axios.get(`/api/v1/product/get-product/${slug}`);
    if (!data.product) {
      toast.error("Item does not exist");
      return;
    }

    const inventory = data.product.quantity;

    if (cart[slug] && cart[slug].quantity + 1 > inventory) {
      toast.error("Error added to cart: Not enough inventory");
      return;
    }

    if (!data.product.price) {
      toast.error("Error added to cart: Price of product not available");
      return;
    }

    dispatch({ type: "ADD_TO_CART", payload: { slug, price: data.product.price, productId: data.product._id } });
  };

  const removeFromCart = (slug) => {
    dispatch({ type: "REMOVE_FROM_CART", payload: { slug } });
  };

  const updateQuantity = async (slug, quantity) => {
    const { data } = await axios.get(`/api/v1/product/get-product/${slug}`);
    if (!data.product) {
      toast.error("Item does not exist");
      return;
    }

    const inventory = data.product.quantity;

    if (cart[slug] && quantity > inventory) {
      toast.error("Error updating quantity: Not enough inventory");
      return;
    }

    if (!data.product.price) {
      toast.error("Error added to cart: Price of product not available");
      return;
    }

    dispatch({ type: "UPDATE_QUANTITY", payload: { slug, quantity, price: data.product.price, productId: data.product._id } });
  };

  const clearCart = () => {
    dispatch({ type: "CLEAR_CART" });
  };

  return (
    <CartContext.Provider
      value={{
        cart,
        addToCart,
        removeFromCart,
        updateQuantity,
        clearCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  return useContext(CartContext);
};
