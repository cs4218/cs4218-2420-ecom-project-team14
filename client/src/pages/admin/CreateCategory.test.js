import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { act } from "react-dom/test-utils"
import "@testing-library/jest-dom";
import CreateCategory from "./CreateCategory";
import { useAuth } from "../../context/auth";
import axios from "axios";
import toast from "react-hot-toast";

jest.mock("../../components/Layout", () => ({ children }) => (
  <div data-testid="mock-layout">{children}</div>
));
jest.mock("../../components/AdminMenu", () => () => (
  <div data-testid="mock-admin-menu">Admin Menu</div>
));
jest.mock("../../components/Form/CategoryForm", () => ({ handleSubmit, value, setValue }) => (
  <div data-testid="mock-category-form">
    <input
      data-testid="category-input"
      value={value}
      onChange={(e) => setValue(e.target.value)}
    />
    <button onClick={handleSubmit}>Submit</button>
  </div>
));

jest.mock("axios");
jest.mock("react-hot-toast");

describe("CreateCategory Component", () => {
    const mockCategories = [
    { _id: "1", name: "Not Shirt" },
    { _id: "2", name: "Shirt" },
    ];

    beforeEach(() => {
        axios.get.mockResolvedValue({ data: { success: true, category: mockCategories } });
        axios.post.mockResolvedValue({ data: { success: true } });
        axios.put.mockResolvedValue({ data: { success: true } });
        axios.delete.mockResolvedValue({ data: { success: true } });
    });

    // Rendering Tests
    describe("Component Rendering", () => {
        it("renders CreateCategory without crashing", async () => {
            await act(async () => {
                render(<CreateCategory />);
            });
            expect(screen.getByTestId("mock-layout")).toBeInTheDocument();
            expect(screen.getByTestId("mock-admin-menu")).toBeInTheDocument();
            expect(screen.getByTestId("mock-category-form")).toBeInTheDocument();
        });

        it("fetches and displays categories", async () => {
            await act(async () => {
                render(<CreateCategory />);
            });
            await waitFor(() => {
                expect(screen.getByText("Not Shirt")).toBeInTheDocument();
            expect(screen.getByText("Shirt")).toBeInTheDocument();
            });
        });
    });

    // API Call Tests
    describe("Fetching Categories", () => {
        it("calls getAllCategory() on mount", async () => {
            await act(async () => {
                render(<CreateCategory />);
            });
            await waitFor(() => {
                expect(axios.get).toHaveBeenCalledWith("/api/v1/category/get-category");
            });
        });
    });

    // Create Category Tests
    describe("Creating a New Category", () => {
        it("submits the category form and calls API", async () => {
            await act(async () => {
                render(<CreateCategory />);
            });

            const input = screen.getByTestId("category-input");
            fireEvent.change(input, { target: { value: "Maybe Shirt" } });
            fireEvent.click(screen.getByText("Submit"));

            await waitFor(() => {
                expect(axios.post).toHaveBeenCalledWith("/api/v1/category/create-category", {
                    name: "Maybe Shirt",
                });
                expect(toast.success).toHaveBeenCalledWith("Maybe Shirt is created");
            });
        });
    });

    // UPdate Category Tests
    describe("Updating a Category", () => {
        it("updates an existing category", async () => {
            await act(async () => {
                render(<CreateCategory />);
            });
            await waitFor(() => {
                expect(screen.getByText("Not Shirt")).toBeInTheDocument();
                expect(screen.getByText("Shirt")).toBeInTheDocument();
            });

            const editButtons = screen.getAllByText("Edit");
            await act(async () => {
                fireEvent.click(editButtons[0]);
            });

            const updatedInput = screen.getByDisplayValue("Not Shirt");

            await act(async () => {
                fireEvent.change(updatedInput, { target: { value: "Definitely Shirt" } });
            });

            const submitButton = screen.getAllByText("Submit");
            await act(async () => {
                fireEvent.click(submitButton[1]); // Brah why are there multiple submit buttons?
                // Whatever, it works fine now.
            });

            await waitFor(() => {
                expect(axios.put).toHaveBeenCalledWith(
                "/api/v1/category/update-category/1",
                { name: "Definitely Shirt" }
                );
                expect(toast.success).toHaveBeenCalledWith("Definitely Shirt is updated");
            });
        });
    });

    // Delete Category Tests
    describe("Deleting a Category", () => {
        it("deletes a category", async () => {
            await act(async () => {
                render(<CreateCategory />);
            });

            await waitFor(() => {
                expect(screen.getByText("Not Shirt")).toBeInTheDocument();
                expect(screen.getByText("Shirt")).toBeInTheDocument();
            });

            const deleteButtons = screen.getAllByText("Delete");
            fireEvent.click(deleteButtons[1]); 

            await waitFor(() => {
                expect(axios.delete).toHaveBeenCalledWith("/api/v1/category/delete-category/2");
                expect(toast.success).toHaveBeenCalledWith("category is deleted");
            });
        });
    });

    // Snapshot Test
    describe("Snapshot", () => {
        it("matches snapshot", async () => {
            axios.get.mockResolvedValueOnce({
                data: { success: true, category: [] },
            });
            let renderResult;
            await act(async () => {
                renderResult = render(<CreateCategory />);
            });
            const { container } = renderResult;
            expect(container).toMatchSnapshot();
        });
      });
});
