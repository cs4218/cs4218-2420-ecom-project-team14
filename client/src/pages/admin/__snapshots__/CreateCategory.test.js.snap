// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCategory Component Snapshot matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-layout"
  >
    <div
      class="container-fluid m-3 p-3"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          <div
            data-testid="mock-admin-menu"
          >
            Admin Menu
          </div>
        </div>
        <div
          class="col-md-9"
        >
          <h1>
            Manage Category
          </h1>
          <div
            class="p-3 w-50"
          >
            <div
              data-testid="mock-category-form"
            >
              <input
                data-testid="category-input"
                value=""
              />
              <button>
                Submit
              </button>
            </div>
          </div>
          <div
            class="w-75"
          >
            <table
              class="table"
            >
              <thead>
                <tr>
                  <th
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    scope="col"
                  >
                    Actions
                  </th>
                </tr>
              </thead>
              <tbody />
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
